src/libsyntax/parse/parser.rs

```diff
@@ -5547,8 +5548,16 @@ impl<'a> Parser<'a> {
 
                 let ty_first = ty_first.into_inner();
                 let path = match ty_first.node {
-                    // This notably includes paths passed through `ty` macro fragments (#46438).
-                    TyKind::Path(None, path) => path,
+                    TyKind::Path(None, path) => {
+                        if let Some(span) = ty_first_interpolated {
+                            self.diagnostic()
+                                .struct_span_warn(span, "expected a trait, found type")
+                                .note("this warning will become a hard error in a future release")
```

Comment:
Is there a reason we can't use the standard lint infrastructure?

Reply:
It's too early in the parser, we don't even have `NodeId`s yet to attach the lint.

Reply:
OK. Before we merge, let's try to [follow the other steps though](https://forge.rust-lang.org/rustc-bug-fix-procedure.html), i.e., create a tracking issue and so forth.